source/mothership-design.md

```diff
@@ -0,0 +1,771 @@
+   spinner delays for a brief period before checking again, and delays after
+   each check until an event triggers the spinner.
+
+[^spinners]: Spinner: Event loop, where events are items (packets or messages)
+    in the consumer queue.
+
+The discrepancy between spinner types is to encourage context to prioritise
+threads with fast spinners. The threads are:
+
+ - `main`: The root thread, which spawns the other threads below and waits for
+   them to exit. All threads are running for the duration of the Mothership
+   process (excepting process startup and teardown).
+
+ - `MPIInputBroker`: Responsible for filtering MPI messages into either the
+   Mothership object C&C queue (`MPICncQueue`), or the Application-handling
+   queue (`MPIApplicationQueue`). This "splitting" is needed because, if a
+   poorly-written supervisor[^malicious] floods with MPI messages, it will not
+   be possible for the operator to stop the application in reasonable time,
+   potentially compromising the running of other applications on the compute
+   fabric. It calls `CommonBase::MPISpinner`. This is a fast spinner.
+
+ - `MPICncResolver`: Responsible for draining the `MPICncQueue` queue, and
+   enacting those messages in sequence as appropriate. This is a slow spinner.
+
+ - `MPIApplicationResolver`: As above, but for the `MPIApplicationQueue`
+   queue. This is a slow spinner. These messages will either be:
+
+   - Converted into packets for the compute fabric, and placed in the
+     `BackendOutputQueue` queue.
+
+   - Used to call a supervisor method (which may in turn produce more
+     traffic).
+
+ - `BackendOutputBroker`: Responsible for draining the `BackendOutputQueue`
+   queue by sending packets into the compute fabric. This is a slow spinner,
+   which spins not only on the existence of packets in the queue, but also on
+   being able to send the message (on asking the compute fabric).
+
+ - `BackendInputBroker`: Responsible for draining the compute fabric into a
+   large (but finite) queue buffer (`BackendInputQueue`)[^backendinput]. When
+   there are no packets to drain from the compute fabric, or when the
+   `BackendInputQueue` buffer is full, this thread converts the next packet
+   into a message, and either sends it to the appropriate destination over MPI,
+   or pushes it to `MPICncQueue` (Tinsel Command and Control). Note that it
+   does not push to `MPIApplicationQueue` (Supervisor) for fairness. This is a
+   fast spinner.
+
+ - `DebugInputBroker`: Responsible for draining the debug fabric into a large
+   queue buffer (`DebugInputQueue`). When there are no packets to drain, or
+   when the `DebugInputQueue` buffer is full, this thread converts the next
+   packet into a message to be `Post`-ed, using MPI. This is a slow spinner.
+
+[^malicious]: Or a maliciously-written supervisor, but we assume people will be
+    playing nice for now.
+
+[^backendinput]: I know Tinsel implements such a queue, but other backends may
```

Comment:
I don't follow; are you suggesting that:

 - All backends implement their own input queue (i.e. packets from the compute fabric to the Mothership), by polling themselves, or

 - The Mothership implements `BackendInputQueue` as described, and no backends implement their own queue.

The former makes sense on the surface, but remember that we (the Orchestrator) would then have no control over the maximum size of the queue, or any failure mode that occurs. We would have to mandate that the size and failure modes are configurable. The latter makes more sense to me, because it devolves responsibility of buffering input messages to the Mothership, as opposed to the backend (which is where we want it). Unfortunately, our only backend so far (Tinsel) has an input queue of sorts.

Reply:
honestly, I am not sure what the correct answer is.

A counter question is why does the Orchestrator need to configure the queue size and failure modes? This is very much a backend-specific "feature" so surely it is up to the backend to have adequate buffer and error handling?

Reply:
It is important for the Orchestrator to have control over the queue size, or at least be able to handle when the queue is full, because then the Orchestrator can send a message with its contents. If we didn't do this, the backend would need to somehow communicate with the Orchestrator to send the message now (because there's no space for anything else).

Reply:
makes sense.

So the question is do we "mandate" that backends have a queue-free config option or a minimum set of configurable parameters?